File: mailblastr-java/src/test/java/com/mailblastr/tests/ContactsTest.java
package com.mailblastr.tests;

import com.mailblastr.Mailblastr;
import com.mailblastr.requests.BatchContactsRequest;
import com.mailblastr.requests.ContactInput;
import com.mailblastr.requests.CreateContactRequest;
import com.mailblastr.requests.ImportContactsRequest;
import com.mailblastr.requests.ListContactsParams;
import com.mailblastr.requests.UpdateContactRequest;
import com.mailblastr.requests.UpdateContactTopicsRequest;

/** Domain-first contacts: flat vs nested routing, imports, segments, topics. */
public final class ContactsTest {
    public static void run() {
        Check.suite("ContactsTest");
        StubTransport t = new StubTransport();
        Mailblastr mb = new Mailblastr("mb_test_key", "https://api.test", t);
        t.respond(200, "{\"object\":\"contact\",\"id\":\"c_1\"}");

        // --- flat create carries domain in the body ---
        mb.contacts().create(CreateContactRequest.builder()
                .domain("example.com")
                .email("ada@lovelace.dev")
                .firstName("Ada")
                .property("plan", "pro")
                .build());
        Check.eq("flat create url", "https://api.test/contacts", t.lastUrl);
        Check.eq("flat create method", "POST", t.lastMethod);
        Check.contains("flat create body has domain", t.lastBody, "\"domain\":\"example.com\"");
        Check.contains("flat create body has properties", t.lastBody, "\"properties\":{\"plan\":\"pro\"}");

        // --- nested create routes via the audience path, no domain in body ---
        mb.contacts().create(CreateContactRequest.builder()
                .audienceId("aud_1")
                .email("ada@lovelace.dev")
                .build());
        Check.eq("nested create url", "https://api.test/audiences/aud_1/contacts", t.lastUrl);
        Check.notContains("nested create body has no domain", t.lastBody, "domain");
        Check.notContains("nested create body has no audienceId", t.lastBody, "audienceId");

        // --- get by email + domain (email is percent-encoded) ---
        mb.contacts().get("ada@lovelace.dev", "example.com");
        Check.eq("get by email url",
                "https://api.test/contacts/ada%40lovelace.dev?domain=example.com", t.lastUrl);
        mb.contacts().get("c_1");
        Check.eq("get by id url", "https://api.test/contacts/c_1", t.lastUrl);
        mb.contacts().getByAudience("aud_1", "c_1");
        Check.eq("get nested url", "https://api.test/audiences/aud_1/contacts/c_1", t.lastUrl);

        // --- list: flat requires domain; nested drops the domain param ---
        mb.contacts().list(ListContactsParams.builder()
                .domain("example.com").limit(10).segmentId("seg_1").build());
        Check.eq("flat list url",
                "https://api.test/contacts?domain=example.com&limit=10&segment_id=seg_1", t.lastUrl);
        mb.contacts().list(ListContactsParams.builder()
                .audienceId("aud_1").domain("example.com").limit(5).build());
        Check.eq("nested list url (no domain param)",
                "https://api.test/audiences/aud_1/contacts?limit=5", t.lastUrl);
        mb.contacts().list("example.com");
        Check.eq("list shortcut url", "https://api.test/contacts?domain=example.com", t.lastUrl);

        // --- batch import ---
        mb.contacts().batch(BatchContactsRequest.builder()
                .audienceId("aud_1")
                .contact(ContactInput.builder().email("a@b.com").firstName("A").build())
                .onConflict("skip")
                .build());
        Check.eq("batch url", "https://api.test/audiences/aud_1/contacts/batch?on_conflict=skip", t.lastUrl);
        Check.eq("batch body",
                "{\"contacts\":[{\"email\":\"a@b.com\",\"first_name\":\"A\"}]}", t.lastBody);

        // --- CSV import with strict properties ---
        mb.contacts().importCsv(ImportContactsRequest.builder()
                .audienceId("aud_1")
                .csv("email,company\na@b.com,Acme")
                .createProperties(false)
                .build());
        Check.eq("import url",
                "https://api.test/audiences/aud_1/contacts/import?create_properties=false", t.lastUrl);
        Check.eq("import body", "{\"csv\":\"email,company\\na@b.com,Acme\"}", t.lastBody);

        // --- update: flat with email id + domain disambiguator ---
        mb.contacts().update(UpdateContactRequest.builder()
                .id("ada@lovelace.dev").domain("example.com").unsubscribed(true).build());
        Check.eq("flat update url", "https://api.test/contacts/ada%40lovelace.dev", t.lastUrl);
        Check.eq("flat update method", "PATCH", t.lastMethod);
        Check.eq("flat update body",
                "{\"unsubscribed\":true,\"domain\":\"example.com\"}", t.lastBody);

        // --- update nested ---
        mb.contacts().update(UpdateContactRequest.builder()
                .audienceId("aud_1").id("c_1").firstName("Ada").build());
        Check.eq("nested update url", "https://api.test/audiences/aud_1/contacts/c_1", t.lastUrl);
        Check.notContains("nested update body has no domain", t.lastBody, "domain");

        // --- remove variants ---
        mb.contacts().remove("ada@lovelace.dev", "example.com");
        Check.eq("flat remove url",
                "https://api.test/contacts/ada%40lovelace.dev?domain=example.com", t.lastUrl);
        Check.eq("remove method", "DELETE", t.lastMethod);
        mb.contacts().removeByAudience("aud_1", "c_1");
        Check.eq("nested remove url", "https://api.test/audiences/aud_1/contacts/c_1", t.lastUrl);

        // --- segment membership + topics ---
        mb.contacts().addToSegment("c_1", "seg_1");
        Check.eq("addToSegment url", "https://api.test/contacts/c_1/segments/seg_1", t.lastUrl);
        Check.eq("addToSegment method", "POST", t.lastMethod);
        mb.contacts().removeFromSegment("c_1", "seg_1");
        Check.eq("removeFromSegment method", "DELETE", t.lastMethod);
        mb.contacts().listSegments("c_1");
        Check.eq("listSegments url", "https://api.test/contacts/c_1/segments", t.lastUrl);
        mb.contacts().getTopics("c_1");
        Check.eq("getTopics url", "https://api.test/contacts/c_1/topics", t.lastUrl);
        mb.contacts().updateTopics("c_1", UpdateContactTopicsRequest.builder()
                .optIn("top_1").optOut("top_2").build());
        Check.eq("updateTopics method", "PATCH", t.lastMethod);
        Check.eq("updateTopics body",
                "{\"topics\":[{\"id\":\"top_1\",\"subscription\":\"opt_in\"},"
                        + "{\"id\":\"top_2\",\"subscription\":\"opt_out\"}]}",
                t.lastBody);
    }

    public static void main(String[] args) {
        run();
        Check.finish();
    }
}

File: mailblastr-php/tests/cases/ContactsTest.php
<?php

declare(strict_types=1);

// ---- create: flat, domain-first ----
[$mb, $t] = make_client();
$mb->contacts->create([
    'domain' => 'example.com',
    'email' => 'user@example.com',
    'first_name' => 'Ada',
]);
check_same('contacts.create flat: method', 'POST', $t->last()['method']);
check_same('contacts.create flat: path', '/contacts', $t->lastPath());
check_same('contacts.create flat: domain kept in body', [
    'domain' => 'example.com',
    'email' => 'user@example.com',
    'first_name' => 'Ada',
], $t->lastJson());

// ---- create: nested audience route strips audienceId + domain from body ----
$mb->contacts->create([
    'audienceId' => 'aud_1',
    'domain' => 'example.com',
    'email' => 'user@example.com',
]);
check_same('contacts.create nested: path', '/audiences/aud_1/contacts', $t->lastPath());
check_same('contacts.create nested: body has neither audienceId nor domain', ['email' => 'user@example.com'], $t->lastJson());

// ---- get: by email + domain ----
$mb->contacts->get(['id' => 'user@example.com', 'domain' => 'example.com']);
check_same('contacts.get: email id encoded + domain query', '/contacts/user%40example.com?domain=example.com', $t->lastPath());

$mb->contacts->get(['id' => 'cont_1', 'audienceId' => 'aud_1']);
check_same('contacts.get nested: path', '/audiences/aud_1/contacts/cont_1', $t->lastPath());

// ---- list: flat requires domain; nested ignores it ----
$mb->contacts->list(['domain' => 'example.com', 'limit' => 10, 'segment_id' => 'seg_1']);
check_same('contacts.list flat: path', '/contacts?domain=example.com&limit=10&segment_id=seg_1', $t->lastPath());

$mb->contacts->list(['audienceId' => 'aud_1', 'domain' => 'example.com', 'limit' => 10]);
check_same('contacts.list nested: domain omitted from query', '/audiences/aud_1/contacts?limit=10', $t->lastPath());

// ---- batch import ----
$mb->contacts->batch([
    'audienceId' => 'aud_1',
    'on_conflict' => 'skip',
    'contacts' => [['email' => 'a@b.com'], ['email' => 'c@d.com']],
]);
check_same('contacts.batch: path', '/audiences/aud_1/contacts/batch?on_conflict=skip', $t->lastPath());
check_same('contacts.batch: body wraps contacts', ['contacts' => [['email' => 'a@b.com'], ['email' => 'c@d.com']]], $t->lastJson());

// ---- CSV import ----
$mb->contacts->import([
    'audienceId' => 'aud_1',
    'csv' => "email,company\na@b.com,Acme",
    'create_properties' => false,
]);
check_same('contacts.import: strict-mode query', '/audiences/aud_1/contacts/import?create_properties=false', $t->lastPath());
check_same('contacts.import: csv body', ['csv' => "email,company\na@b.com,Acme"], $t->lastJson());

// segment_id rides on the query string alongside on_conflict
$mb->contacts->import([
    'audienceId' => 'aud_1',
    'csv' => 'email\na@b.com',
    'file_name' => 'leads.csv',
    'on_conflict' => 'skip',
    'segment_id' => 'seg_1',
]);
check_same('contacts.import: segment_id query', '/audiences/aud_1/contacts/import?on_conflict=skip&segment_id=seg_1', $t->lastPath());
check_same('contacts.import: file_name body', ['csv' => 'email\na@b.com', 'file_name' => 'leads.csv'], $t->lastJson());

// pre-uploaded file mode sends storage_key instead of csv
$mb->contacts->import(['audienceId' => 'aud_1', 'storage_key' => 'imports/abc.csv']);
check_same('contacts.import: storage_key body', ['storage_key' => 'imports/abc.csv'], $t->lastJson());

// ---- presigned direct upload ----
$mb->contacts->uploadUrl(['audienceId' => 'aud_1', 'filename' => 'leads.csv', 'size' => 2048]);
check_same('contacts.uploadUrl: method', 'POST', $t->last()['method']);
check_same('contacts.uploadUrl: path', '/audiences/aud_1/contacts/import/upload', $t->lastPath());
check_same('contacts.uploadUrl: body', ['filename' => 'leads.csv', 'size' => 2048], $t->lastJson());

// ---- update: flat by email keeps domain in body ----
$mb->contacts->update([
    'id' => 'user@example.com',
    'domain' => 'example.com',
    'unsubscribed' => true,
]);
check_same('contacts.update flat: method', 'PATCH', $t->last()['method']);
check_same('contacts.update flat: path', '/contacts/user%40example.com', $t->lastPath());
check_same('contacts.update flat: body', ['domain' => 'example.com', 'unsubscribed' => true], $t->lastJson());

// ---- remove: nested + flat with domain ----
$mb->contacts->remove(['id' => 'cont_1', 'audienceId' => 'aud_1']);
check_same('contacts.remove nested: method', 'DELETE', $t->last()['method']);
check_same('contacts.remove nested: path', '/audiences/aud_1/contacts/cont_1', $t->lastPath());

$mb->contacts->remove(['id' => 'user@example.com', 'domain' => 'example.com']);
check_same('contacts.remove flat: path', '/contacts/user%40example.com?domain=example.com', $t->lastPath());

// ---- segments + topics helpers ----
$mb->contacts->addToSegment('cont_1', 'seg_1');
check_same('contacts.addToSegment: method', 'POST', $t->last()['method']);
check_same('contacts.addToSegment: path', '/contacts/cont_1/segments/seg_1', $t->lastPath());

$mb->contacts->removeFromSegment('cont_1', 'seg_1');
check_same('contacts.removeFromSegment: method', 'DELETE', $t->last()['method']);

$mb->contacts->listSegments('cont_1');
check_same('contacts.listSegments: path', '/contacts/cont_1/segments', $t->lastPath());
$mb->contacts->listSegments('cont_1', ['limit' => 5]);
check_same('contacts.listSegments: pagination', '/contacts/cont_1/segments?limit=5', $t->lastPath());

$mb->contacts->getTopics('cont_1');
check_same('contacts.getTopics: path', '/contacts/cont_1/topics', $t->lastPath());
$mb->contacts->getTopics('cont_1', ['limit' => 5]);
check_same('contacts.getTopics: pagination', '/contacts/cont_1/topics?limit=5', $t->lastPath());

$mb->contacts->updateTopics('cont_1', ['topics' => [['id' => 'top_1', 'subscription' => 'opt_in']]]);
check_same('contacts.updateTopics: method', 'PATCH', $t->last()['method']);
check_same('contacts.updateTopics: path', '/contacts/cont_1/topics', $t->lastPath());
check_same('contacts.updateTopics: body', ['topics' => [['id' => 'top_1', 'subscription' => 'opt_in']]], $t->lastJson());
